### DISCUSSION TITLE
Proposal: Support for "where blocks" inside methods

### DISCUSSION THREAD

**Discussion by lskyum** [dotnet/csharplang#6943] [Language Ideas] [answered]
The proposal is to support "where blocks" inside methods, so types can be checked without dynamic type checking.

For example for this implementation of the method INumberBase.TryConvertFromChecked:
```
public static bool TryConvertFromChecked<TOther>(TOther value, out Fraction result) where TOther : INumberBase<TOther>
{
	where(TOther : double)
	{
		// Here TOther is known to be double and any use of TOther will be as if it's a double:
		double x = value;  // Compiles inside the where block, but not outside
	}
	where(TOther : decimal)
	{
		// Here TOther is known to be decimal...
	}
}
```

Where dynamic type checking is something like this: 
```
public static bool TryConvertFromChecked<TOther>(TOther value, out Fraction result) where TOther : INumberBase<TOther>
{
	if(typeof(TOther) == typeof(double))
	{
		// Here TOther is known to be double, but values needs casting anyway
		double x = (double)(object)value;
	}
	if(typeof(TOther) == typeof(decimal))
	{
	}
}
```
... or perhaps a switch statement, but either way, it will be less efficient and more clunky than a static check.

Another benefit is that static checks will be more friendly towards AOT and trimming, which may actually turn out to be the biggest benefit.

**Comment by HaloFour** [comment, marked as answer]
See: #6308

Note that the runtime explicitly looks for the pattern of `if(typeof(TOther) == typeof(double))` and can use that to specialize the method and remove the type checks entirely.
